tests/test_odp_event_manager.py

```diff
@@ -0,0 +1,434 @@
+
+        mock_send.assert_called_once_with(self.api_key, self.api_host, self.processed_events)
+        self.assertEqual(len(event_manager._current_batch), 2)
+        mock_logger.debug.assert_any_call('Error dispatching ODP events, scheduled to retry.')
+
+        mock_logger.reset_mock()
+
+        with mock.patch.object(
+            event_manager.zaius_manager, 'send_odp_events', new_callable=CopyingMock, return_value=False
+        ) as mock_send:
+            event_manager.stop()
+
+        mock_send.assert_called_once_with(self.api_key, self.api_host, self.processed_events)
+        self.assertEqual(len(event_manager._current_batch), 0)
+        mock_logger.error.assert_not_called()
+
+    def test_odp_event_manager_send_failure(self):
+        mock_logger = mock.Mock()
+        event_manager = OdpEventManager(self.odp_config, mock_logger)
+        event_manager.start()
+
+        with mock.patch.object(
+            event_manager.zaius_manager,
+            'send_odp_events',
+            new_callable=CopyingMock,
+            side_effect=Exception('Unexpected error')
+        ) as mock_send, mock.patch('uuid.uuid4', return_value=self.test_uuid):
+            event_manager.send_event(**self.events[0])
+            event_manager.send_event(**self.events[1])
+            event_manager.flush()
+            event_manager.event_queue.join()
+
+        mock_send.assert_called_once_with(self.api_key, self.api_host, self.processed_events)
+        self.assertEqual(len(event_manager._current_batch), 0)
+        mock_logger.error.assert_any_call(f"ODP event send failed ({self.processed_events} Unexpected error).")
+        self.assertStrictTrue(event_manager.is_running)
+        event_manager.stop()
+
+    def test_odp_event_manager_disabled(self):
+        mock_logger = mock.Mock()
+        odp_config = OdpConfig()
+        odp_config.update(None, None, None)
+        event_manager = OdpEventManager(odp_config, mock_logger)
+        event_manager.start()
+
+        event_manager.send_event(**self.events[0])
+        event_manager.send_event(**self.events[1])
+        event_manager.event_queue.join()
+
+        self.assertEqual(len(event_manager._current_batch), 0)
+        mock_logger.error.assert_not_called()
+        mock_logger.debug.assert_any_call('ODP event queue has been disabled.')
+        self.assertStrictTrue(event_manager.is_running)
+        event_manager.stop()
+
+    def test_odp_event_manager_queue_full(self):
+        mock_logger = mock.Mock()
+
+        with mock.patch('optimizely.helpers.enums.OdpEventManagerConfig.DEFAULT_QUEUE_CAPACITY', 1):
+            event_manager = OdpEventManager(self.odp_config, mock_logger)
+
+        with mock.patch('optimizely.odp.odp_event_manager.OdpEventManager.is_running', True):
+            event_manager.send_event(**self.events[0])
+            event_manager.send_event(**self.events[1])
+
+        mock_logger.warning.assert_called_once_with('ODP event send failed (Queue is full).')
```

Comment:
Is this "warning" (not "error") a good level? Can you check if this is consistent with logx event overflow?

Reply:
This is consistent with logx event overflow. However since the event is dropped, error level doesn't seem out of place, in my opinion.

Reply:
@andrewleap-optimizely if it's consistent with logx overflow, let's keep it consistent. error or warning is debatable since overflow may be a result of other errors.